Handle MAV_CMD_DO_CHANGE_SPEED command for FW Offboard global positio…

ThomasRigi opened the pull request:
…n control.

Please use [PX4 Discuss](http://discuss.px4.io/) or [Slack](http://slack.px4.io/) to align on pull requests if necessary. You can then open draft pull requests to get early feedback.
Initial slack conversation with @dagar : https://px4.slack.com/archives/C7L8W6DAL/p1593439308005400

Maybe @Jaeyoung-Lim who has added support for SET_POSITION_TARGET_GLOBAL_INT in the first place is also interested in this PR :)

**Describe problem solved by this pull request**
In FW Offboard control using SET_POSITION_TARGET_GLOBAL_INT the command  MAV_CMD_DO_CHANGE_SPEED had no effect. This PR adds support for this without affecting any existing behaviour.

**Describe your solution**
The MavlinkReceiver class handles MAV_CMD_DO_CHANGE_SPEED commands by saving the desired speed to a member variable based on whether the drone is in FW or in MC mode. This speed is then accessed in `MavlinkReceiver::handle_message_set_position_target_global_int(mavlink_message_t *msg)` when position control is active and published in the `pos_sp_triplet.current.cruising_speed` field.

Upon a flight mode change, the cruising speed values are reset to default. 

**Describe possible alternatives**
I had also briefly considered an option where the cruising speed would only be set for the flight mode in which the command is received and would stick in this mode even when switching to a different mode and back (e.g. offboard --> loiter --> offboard). I decided against this approach as there might be unwanted old cruising speeds lingering around.

See Additional context for more details.

**Test data / coverage**
Gazebo standard vtol:

1. Two change speed commands in offboard mode: first one in NW corner to 18m/s, second one back to default before back transition. : https://logs.px4.io/plot_app?log=17171469-7155-4f56-b9fe-4541070a8ffb

2. Change speed to 18m/s in NW corner, followed by Loiter which resets to default. Then continuing in offboard, showing that the old change speed command is indeed forgotten: https://logs.px4.io/plot_app?log=d15040f0-3ea1-4ccf-8efa-494b32f02ef0

3. Change speed to 18m/s in NW corner, followed by RTL which resets to default. https://logs.px4.io/plot_app?log=726f8af1-2a2b-462c-a8c2-7f536574ca51

4. Change speed in Mission, FW mode: behaviour unchanged. https://logs.px4.io/plot_app?log=8e5d2ee9-2ed1-4922-8e91-693be4411ee1

5. Change speed in Mission, MC mode: behaviour unchanged. https://logs.px4.io/plot_app?log=4ed13e70-d137-4c2a-9f4d-7743f54ae0ab


**Additional context**
As said in the description, this PR only adds support in FW offboard mode without affecting existing behaviour. This was a choice on purpose, but in principle the approach could also be used for all other modes for which the MAV_CMD_DO_CHANGE_SPEED command could make sense. You would just have to publish the desired cruising speed in a uorb topic to which the different flight mode modules subscribe. 

Going through all PX4 flightmodes, the following are the ones in which IMO it could make sense to have MAV_CMD_DO_CHANGE_SPEED supported:

- Offboard Position control:
1. Our usecase (this PR): change speed in FW offboard position control using SET_POSITION_TARGET_GLOBAL_INT
2. same, but for MC: With this PR the desired MC speed is also published in the `pos_sp_triplet.current.cruising_speed` field, but it's not of much use as the offboard flight task doesn't directly expose a way to set this. It would need to be passed to https://github.com/PX4/Firmware/blob/master/src/modules/mc_pos_control/mc_pos_control_main.cpp#L369 .

- Mission:
Already handled in the navigator module. Architecturally it might make more sense to unify where  MAV_CMD_DO_CHANGE_SPEED is handled (in MavlinkReceiver) and then access this from the navigator (if the decision is taken to add support for the change speed command for multiple flight modes).

- RTL:
The main reason why I decided to reset the cruising speed to default after every mode change. In (the relatively unlikely) case that a bad change speed command has put the drone in a state that requires an RTL action then this bad cruising speed should certainly not stick around.
I currently don't know exactly where the speed is set in RTL mode, but I imagine that with a uorb topic published from MavlinkReceiver it should be relatively easy to have this working.


- FW Loiter:
Would be easy to implement with the uorb topic.

- Orbit:
A mode that we have never used and probably never will on our VTOL drone, but I guess if you set the cruising speed as a limit in the MC Position control module it should also apply to orbit.



We currently only need MAV_CMD_DO_CHANGE_SPEED for FW offboard and would be really happy to have this merged into master without having to sort out all the details about the other flight modes first. In case there is enough positive feedback to also implement it for the other flight modes, I would prefer to have a separate follow-up PR for that.

Any inputs are welcome :)

dayjaby commented:
I think navigator_main.cpp is the better place to handle cruising speed changes. The disadvantage of handling certain commands in the mavlink receiver is that those commands are not handled correctly if they occur in a mission instead of being sent as mavlink command.

ThomasRigi commented:
@dayjaby That's a very valid point, thanks for pointing it out.

So then the navigator should publish the cruising speed in a uorb topic which is then read in the FW and MC position control modules for all relevant flight modes. I think the logic that I outlined above still applies: upon a mode switch the cruising speed should get reset to default.

BazookaJoe1900 commented:
can _'Support MAV_CMD_DO_CHANGE_SPEED in auto-loiter #5122'_ be solved too under this pr? which is more relevant for MC

ThomasRigi commented:
@BazookaJoe1900 Not in its current state. Right now it only adds support for FW offboard mode. (but I have my doubts if it is going to be merged like this, as there seems to be a general need to improve how the MAV_CMD_DO_CHANGE_SPEED command is handled)

With what I have outlined in the Additional Context section yes, that approach could include #5122. The goal would be to have one instance (in the Navigator most likely) which is responsible for setting (and maybe keeping track of) the current cruising speed for all flight modes and then publishing it so that the other modules can access it. I had forgotten to list it above as I thought that MC loiter is hovering on the spot only (no speed), but as it's also active during a reposition it makes total sense to also have it supported.

Jaeyoung-Lim commented on src/modules/mavlink/mavlink_receiver.cpp:
```diff
@@ -2818,6 +2822,15 @@ MavlinkReceiver::Run()
+				reset_cruising_speed();
+				_last_nav_state = _vehicle_status.nav_state;
+			}
+		}
```
I think this makes sense if you are exiting offboard modes, but would be quite confusing if you are going through different flight modes within normal flight modes (e.g. stabilized/position)

IMHO, I would rather leave this and make it more transparent. Offboard application on the companion computer should be responsible on ensuring that this is set to a safe value.

sfuhrer replied:
@Jaeyoung-Lim Is this comment still valid from your side? It really only affects offboard mode, no?

Jaeyoung-Lim reviewed (commented):
Thanks, this certainly makes sense and is one of the things that was missing for offboard position control for FW.

I am a bit concerned that the resetting behavior will change the overall behavior and not just offboard mode.

ThomasRigi commented:
Thanks @Jaeyoung-Lim for your input and sorry for the late response! I'll have more time for PX4 in the next weeks.

To address your concerns about the resetting behaviour: this only affects the cruise speed that is set in offboard mode and does not influence other flight modes. I have updated the naming to make it clear. 

The only time these lines have an effect is when you switch out of offboard mode and re-enter into offboard mode later on. I thought that it was better to not have old speed setpoints lingering around when re-entering offboard mode. I admit that you could just as well argue that it's better if the drone has a memory of what speed it was supposed to be flying at in offboard mode. I can delete the reset if you prefer.

I have also rebased on current master. There seems to be some sort of issue where the drone just starts to circle endlessly, which it didn't on v1.11.1 with the exact same offboard script. Change speed command is working. https://logs.px4.io/plot_app?log=468226c7-a90f-465f-bb8e-f7e7a9f08eb2

ThomasRigi commented:
@Jaeyoung-Lim any inputs from your side? I'd really like to get this merged as it helps us shorten the back transition if we can already reduce the speed in FW mode before starting the transition. 

If for you the reset is a no-go I'll take it out. I can easily live with both options.

Thanks!

Jaeyoung-Lim commented:
@ThomasRigi Thanks for the reminder. I would still like to get @sfuhrer 's review on this pull request before it gets in.

sfuhrer commented:
@ThomasRigi @Jaeyoung-Lim  I've personally never used offboard mode and the interaction between mavlink_receiver, navigator and the MC/FW Position controller is really frying my brain.. The changes here seem to be constrained to offboard mode, and I see your point that having it reset when switching in and out. If for your specific application you could also live without the resets and e.g. show that in an RTL the speed is properly reset then I personally would reconsider removing the resets - that would strongly reduce the diff/added code of this PR.

ThomasRigi commented:
@sfuhrer @Jaeyoung-Lim Thank you for your feedback. I can live without the reset, and it's indeed true that without it there are fewer lines of code. Also, I have made it more explicit (line 520 in mavlink_receiver.cpp) that it only affects offboard mode. Also, I added support for position control via set_position_target_local_ned. I don't have an offboard script to test it. But in the end both the position_target_local_ned and position_target_global_int both publish the "same" position setpoint triplet, so there's no reason why one should work and not the other.

Here's a log where you can see the changed target airspeed (to 22m/s) just after 5min: https://logs.px4.io/plot_app?log=623753dd-852f-4f97-b90e-eddfb74ad122

![image](https://user-images.githubusercontent.com/48206725/100444595-f4d3e380-30ab-11eb-9d4e-8e5a4b794ec7.png)

sfuhrer commented on src/modules/mavlink/mavlink_receiver.cpp:
```diff
@@ -516,6 +516,12 @@ void MavlinkReceiver::handle_message_command_both(mavlink_message_t *msg, const
 			_mavlink->request_stop_ulog_streaming();
+
+		} else if (cmd_mavlink.command == MAV_CMD_DO_CHANGE_SPEED) {
+			if (_vehicle_status.nav_state == vehicle_status_s::NAVIGATION_STATE_OFFBOARD) {
```
Offboard-noob question here: why is MAV_CMD_DO_CHANGE_SPEED the only command for which we need to check the current nav_state first if really in offboard mode before applying it?

ThomasRigi replied:
If you don't check if you're in offboard mode before calling set_offb_cruising_speed it would also change the cruise speed for offboard mode even if you sent a change speed command when in another flight mode. Which could lead to unexpected behaviour if you switch to offboard later on.

sfuhrer replied:
Okay, and how is it handled for other commands, like position setpoints? I guess the same is true there, you don't want them to have effect unless you're in offboard no?

ThomasRigi replied:
For offboard mode the position setpoints aren't sent as one-time commands, but as a stream of messages. Which are then handled in e.g. https://github.com/PX4/PX4-Autopilot/blob/378b41af6a91cb5dfcabb24421f0a0c9e21a98e5/src/modules/mavlink/mavlink_receiver.cpp#L974. Which in turn checks if offboard is enabled: https://github.com/PX4/PX4-Autopilot/blob/378b41af6a91cb5dfcabb24421f0a0c9e21a98e5/src/modules/mavlink/mavlink_receiver.cpp#L1035. 

Actually, if it makes more sense I could rewrite my code to use the same structure: 
`vehicle_control_mode_s control_mode{};
			_control_mode_sub.copy(&control_mode);

			if (control_mode.flag_control_offboard_enabled) {
                                 set_offb_cruising_speed(cmd_mavlink.param2)
                        }`

Previously with the automatic resets I needed to track the nav_state, but that is no longer true. So it might be more efficient to use _control_mode_sub.copy and get rid of _vehicle_status.

sfuhrer replied:
>For offboard mode the position setpoints aren't sent as one-time commands, but as a stream of messages. Which are then handled in e.g.

Ah yes that makes sense.

>Previously with the automatic resets I needed to track the nav_state, but that is no longer true. So it might be more efficient to use _control_mode_sub.copy and get rid of _vehicle_status.

That I was also wondering. Would further reduce the complexity and diff to current state.

ThomasRigi replied:
Ok, I removed the _vehicle_status. It's true that the diff looks nicer now :) 

One thing I overlooked previously is that I need to check the vehicle type (FW or MC) in the set and get functions, so I need to use 
```
vehicle_status_s vehicle_status{};
_vehicle_status_sub.copy(&vehicle_status);
```
instead inside these functions. Not sure if it's better for the computational performance in offboard mode, but it definitely doesn't affect any other mode now.

sfuhrer replied:
Yes, looks nicely contained now.

sfuhrer commented on src/modules/mavlink/mavlink_receiver.cpp:
```diff
@@ -3101,3 +3112,41 @@ MavlinkReceiver::receive_start(pthread_t *thread, Mavlink *parent)
+	vehicle_status_s vehicle_status{};
+	_vehicle_status_sub.copy(&vehicle_status);
+
+	if (vehicle_status.vehicle_type == vehicle_status_s::VEHICLE_TYPE_ROTARY_WING) {
```
@ThomasRigi this could be compressed a bit. How about 
```suggestion
	if (vehicle_status.vehicle_type == vehicle_status_s::VEHICLE_TYPE_ROTARY_WING && _offb_cruising_speed_mc > 0.0f) {
	return _offb_cruising_speed_mc;
	} else if (vehicle_status.vehicle_type == vehicle_status_s::VEHICLE_TYPE_FIXED_WING && _offb_cruising_speed_fw > 0.0f) {
	return _offb_cruising_speed_fw;
	} else {
	return -1.0f;
	}
```

ThomasRigi replied:
Thanks for the suggestion!

sfuhrer commented:
@ThomasRigi sorry, my suggestion was not meant to be committed 1:1, it doesn't have correct syntax :sweat_smile: .

ThomasRigi commented:
@sfuhrer that explains all the emails about build failures :P Note to myself: don't use the graphical interface on github :sweat_smile: Hope now it's good.

sfuhrer commented:
> @sfuhrer that explains all the emails about build failures :P Note to myself: don't use the graphical interface on github sweat_smile Hope now it's good.

Thanks, looks better that way ;) 

@Jaeyoung-Lim can you have a last look? Then we could bring this in in my opinion.

Jaeyoung-Lim reviewed (approved):
@sfuhrer Thanks, LGTM

@ThomasRigi Would you be able to write this in the documentation of http://docs.px4.io/master/en/flight_modes/offboard.html ?

ThomasRigi commented:
Thanks for merging! :) 

PR for docs: https://github.com/PX4/PX4-user_guide/pull/968
